keps/sig-storage/20190530-pv-health-monitor.md

```diff
@@ -0,0 +1,575 @@
+
+With health monitoring, unhealthy volumes can be detected and fixed early and therefore could prevent more serious problems to occur. While some problems may be corrected automatically by a Kubernetes controller, most problems may involve manual admin intervention or need to be fixed with specific application knowledge. In any case, volume health monitoring is a very important tool that would be beneficial to Kubernetes users.
+
+## User Experience
+
+### Use Cases
+
+Many things could happen to the underlying storage system after a volume is provisioned in Kubernetes.
+
+* Volume could be deleted by accident outside of Kubernetes.
+* The disk that the volume resides on could be removed temporarily for maintenance.
+* The disk that the volume resides on could fail.
+* There may be configuration issues with the underlying storage system that affect the volume’s health.
+* Volume could be out of capacity.
+* The disk may be degrading which affects its performance.
+
+If the volume is mounted on a pod and used by an application, the following problems could also happen.
+* There may be read/write I/O errors.
+* The file system on the volume may be corrupted.
+* Filesystem may be out of capacity.
+* Volume may be unmounted by accident outside of Kubernetes.
+
+If the CSI driver has implemented the CSI volume health function proposed in this design document, Kubernetes could communicate with the CSI driver to retrieve any errors detected by the underlying storage system. Kubernetes can report an event and log an error about this PVC so that user can inquire this information and decide how to handle them. For example, if the volume is out of capacity, user can request a volume expansion to get more space. In the first phase, volume health monitoring is informational only as it is only reported in the events and logs. In the future, we will also look into how Kubernetes may use volume health information to automatically reconcile.
+
+There could be conditions that cannot be reported by a CSI driver. There could be network failure where Kubernetes may not be able to get response from the CSI driver. In this case, a call to the CSI driver may time out. There could be network congestion which causes slow response. One or more nodes where the volume is attached to may be down. This can be monitored and detected by the volume health controller so that user knows what has happened.
```

Comment:
I said "conditions that cannot be reported by a CSI driver" on line 83 so I meant CSI driver cannot report a network failure.  Maybe I should just remove it to avoid confusion.